File: device-platform/frontend/web-console/src/lib/frameProtocol.ts
// Wire protocol: each WebSocket binary message is exactly one frame
//   [1 byte type][payload bytes...]
// Mirrors the backend FrameType enum.

export const FrameType = {
  VIDEO_KEYFRAME:       0x01,
  VIDEO_DELTA:          0x02,
  CONTROL_COMMAND:      0x03,
  INSPECT_REQUEST:      0x04,
  INSPECT_RESPONSE:     0x05,
  HEARTBEAT:            0x06,
  STREAM_METADATA:      0x07,
  FORCE_KEYFRAME:       0x08,
  SCREENSHOT_REQUEST:   0x09,
  SCREENSHOT_RESPONSE:  0x0a,
  // Faz 2 — APK install + app launch + device reset
  APP_INFO_REQUEST:     0x0b,
  APP_INFO_RESPONSE:    0x0c,
  INSTALL_APK_REQUEST:  0x0d,
  INSTALL_APK_RESPONSE: 0x0e,
  LAUNCH_APP_REQUEST:   0x0f,
  LAUNCH_APP_RESPONSE:  0x10,
  RESET_HOME_REQUEST:   0x11,
  RESET_HOME_RESPONSE:  0x12,
} as const;
export type FrameTypeValue = (typeof FrameType)[keyof typeof FrameType];

export type Frame = { type: number; payload: Uint8Array };

export function encodeFrame(type: number, payload: Uint8Array = new Uint8Array(0)): ArrayBuffer {
  const out = new Uint8Array(1 + payload.byteLength);
  out[0] = type & 0xff;
  if (payload.byteLength > 0) out.set(payload, 1);
  return out.buffer;
}

export function decodeFrame(buf: ArrayBuffer): Frame {
  const v = new Uint8Array(buf);
  if (v.byteLength === 0) throw new Error("empty frame");
  return { type: v[0], payload: v.subarray(1) };
}

export function payloadAsString(payload: Uint8Array): string {
  return new TextDecoder().decode(payload);
}

export function jsonFrame(type: number, obj: unknown): ArrayBuffer {
  return encodeFrame(type, new TextEncoder().encode(JSON.stringify(obj)));
}

File: device-platform/products/android/agent/android-agent/app/src/main/kotlin/com/devicefarm/agent/net/FrameProtocol.kt
package com.devicefarm.agent.net

import okio.Buffer
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.nio.charset.StandardCharsets

/**
 * Wire protocol: each WebSocket binary message is exactly one frame.
 *   [1 byte type][payload bytes]
 */
object FrameType {
    const val VIDEO_KEYFRAME: Byte      = 0x01
    const val VIDEO_DELTA:    Byte      = 0x02
    const val CONTROL_COMMAND: Byte     = 0x03
    const val INSPECT_REQUEST: Byte     = 0x04
    const val INSPECT_RESPONSE: Byte    = 0x05
    const val HEARTBEAT: Byte           = 0x06
    const val STREAM_METADATA: Byte     = 0x07
    const val FORCE_KEYFRAME: Byte      = 0x08
    /** hub → agent (JSON: {requestId}) */
    const val SCREENSHOT_REQUEST: Byte  = 0x09
    /** agent → hub: [4 bytes BE metaLen][JSON meta][PNG bytes] */
    const val SCREENSHOT_RESPONSE: Byte = 0x0A

    // ── Faz 2: APK install + app launch + device reset ──────────────────────
    /** hub → agent (JSON: {requestId, packageName}) */
    const val APP_INFO_REQUEST: Byte     = 0x0B
    /** agent → hub (JSON: {requestId, installed, versionCode?, versionName?}) */
    const val APP_INFO_RESPONSE: Byte    = 0x0C
    /** hub → agent (JSON: {requestId, downloadUrl, sha256, expectedVersionCode, packageName}) */
    const val INSTALL_APK_REQUEST: Byte  = 0x0D
    /** agent → hub (JSON: {requestId, status, installedVersionCode?, errorCode?, errorMessage?}) */
    const val INSTALL_APK_RESPONSE: Byte = 0x0E
    /** hub → agent (JSON: {requestId, packageName}) */
    const val LAUNCH_APP_REQUEST: Byte   = 0x0F
    /** agent → hub (JSON: {requestId, status, errorMessage?}) */
    const val LAUNCH_APP_RESPONSE: Byte  = 0x10
    /** hub → agent (JSON: {requestId, packageName?, killProcess?}) */
    const val RESET_HOME_REQUEST: Byte   = 0x11
    /** agent → hub (JSON: {requestId, status, errorMessage?}) */
    const val RESET_HOME_RESPONSE: Byte  = 0x12
}

data class Frame(val type: Byte, val payload: ByteArray) {

    fun encode(): ByteString {
        val buf = Buffer().apply {
            writeByte(type.toInt())
            if (payload.isNotEmpty()) write(payload)
        }
        return buf.readByteString()
    }

    fun payloadAsString(): String = String(payload, StandardCharsets.UTF_8)

    companion object {
        fun decode(bytes: ByteString): Frame {
            val arr = bytes.toByteArray()
            if (arr.isEmpty()) throw IllegalArgumentException("empty frame")
            val type = arr[0]
            val payload = if (arr.size == 1) ByteArray(0) else arr.copyOfRange(1, arr.size)
            return Frame(type, payload)
        }

        fun ofJson(type: Byte, json: String): Frame =
            Frame(type, json.toByteArray(StandardCharsets.UTF_8))

        fun empty(type: Byte): Frame = Frame(type, ByteArray(0))
    }
}

// helper for raw byte arrays
fun ByteArray.toBs(): ByteString = this.toByteString(0, this.size)
